Introduce snapshot testing

philippfromme opened the pull request:
### Proposed Changes

Allows us to easily compare before and after for the defined test cases. Could be added to the CI so the CI only passes once all snapshots pass meaning you either haven't changed the algorithm or you did and you updated the snapshots.

* snapshot tests are part of `npm test` and a snapshot that is not matching will result in an unsuccessful test run
* snapshots that are matching are highlighted in green
* snapshots that are not matching are highlighted in red
* if no snapshot exists no highlight is added

![brave_jawCRplABm](https://github.com/user-attachments/assets/909bc9ea-7205-4c99-bb61-838036d82606)


<!--

Add relevant context (issue fixed or related to), 
a capture of the UI changes (if any) as well as 
steps to try out your changes.

--> 

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] **Brief textual description** of the changes present
* [x] **Visual demo** attached
* [ ] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [ ] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--

Thanks for creating this pull request! ❤️

-->

nikku commented:
Let's get this incorporated soonish, in order to test the other open PRs.

philippfromme commented:
I'll get this PR ready.

barmac commented:
@philippfromme looks that we need to make sure our snapshots are initially valid.

philippfromme commented:
@barmac Thought I'd done that 🙈

barmac commented on test/LayoutSpec.js:
```diff
@@ -1,156 +1,104 @@
+        }
 
-    const generated = generatedContents.map((contents, idx) => {
+        fs.writeFileSync(path.join(outputDirectory, fileName), output, 'utf8');
```
Note that the result is not written if the test fails. I'd suggest to write it before the assertion.

philippfromme replied:
Good point!

barmac commented on test/LayoutSpec.js:
```diff
@@ -1,156 +1,104 @@
-
-        containerParent.appendChild(containerBefore);
-        containerParent.appendChild(containerAfter);
+  after(() => {
```
Quite unexpected to me to see the teardown code on the bottom. Usual practice is to have tests setup on the top, and place test cases later:

before
after
it

barmac commented on test/LayoutSpec.js:
```diff
@@ -1,156 +1,104 @@
-        generatedContents: generatedContents[idx],
-        originalContents: originalContents[idx]
-      };
+        if (UPDATE_SNAPSHOTS) {
```
Does it work given the assertion above?

abdul99ahad commented:
Some of the test cases are not showing in the left panel. 
![image](https://github.com/user-attachments/assets/287f354a-d2b6-4bb1-a6d4-2fb36f237fb9)

philippfromme commented:
> Some of the test cases are not showing in the left panel.

This is because the original diagram doesn't contain any graphical information. While the use case of the library is to create missing graphical information for the snapshot tests it's actually more convinient to have a diagram with graphical information. The library will then simply replace that. I'll update the diagrams that have no DI.

philippfromme commented:
I updated the diagrams without DI and also added an error indication if the import fails completely. Should be clearer now.

![brave_kVlMfV6NoR](https://github.com/user-attachments/assets/5760d4ad-fc52-421e-b8bd-072b891adf93)

philippfromme commented:
Ready for review.

abdul99ahad commented:
> Ready for review.

Looks good to me 🚀

nikku commented:
Updated with a minor hint regarding test updates:

![image](https://github.com/user-attachments/assets/dd6423fe-046b-4e34-9445-c853d57ec610)

nikku reviewed (approved):
Good stuff! I'll look into simplifying our existing tests in a minute.
